gluster/fs/pkg/volume/provision.go

```diff
@@ -0,0 +1,278 @@
+	allocator  gidallocator.Allocator
+}
+
+type glusterBrick struct {
+	Host string
+	Path string
+}
+
+var _ controller.Provisioner = &glusterfsProvisioner{}
+
+func (p *glusterfsProvisioner) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
+	if options.PVC.Spec.Selector != nil {
+		return nil, fmt.Errorf("claim Selector is not supported")
+	}
+	glog.V(4).Infof("Start Provisioning volume: VolumeOptions %v", options)
+
+	gid, err := p.allocator.AllocateNext(options)
+	if err != nil {
+		return nil, err
+	}
+
+	pvcNamespace := options.PVC.Namespace
+	pvcName := options.PVC.Name
+	cfg, err := NewProvisionerConfig(options.PVName, options.Parameters)
+	if err != nil {
+		return nil, fmt.Errorf("Parameter is invalid: %s", err)
+	}
+
+	r, err := p.createVolume(pvcNamespace, pvcName, cfg, gid)
+	if err != nil {
+		return nil, err
+	}
+
+	annotations := make(map[string]string)
+	annotations[annCreatedBy] = createdBy
+	annotations[gidallocator.VolumeGidAnnotationKey] = strconv.FormatInt(int64(gid), 10)
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        options.PVName,
+			Annotations: annotations,
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeReclaimPolicy: options.PersistentVolumeReclaimPolicy,
+			AccessModes:                   options.PVC.Spec.AccessModes,
+			Capacity: v1.ResourceList{
+				v1.ResourceName(v1.ResourceStorage): options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)],
+			},
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				Glusterfs: r,
+			},
+		},
+	}
+	return pv, nil
+}
+
+func (p *glusterfsProvisioner) getClusterNodes(cfg *ProvisionerConfig) []string {
+	// XXX: Improve to get all cluster nodes
+	nodes := make([]string, len(cfg.BrickRootPaths))
+	for i, root := range cfg.BrickRootPaths {
+		nodes[i] = root.Host
+	}
+	return nodes
+}
+
+func (p *glusterfsProvisioner) createVolume(
+	namespace string, name string, cfg *ProvisionerConfig, gid int) (*v1.GlusterfsVolumeSource, error) {
+	var err error
+	var bricks []glusterBrick
+	var endpoint *v1.Endpoints
+	var service *v1.Service
+
+	bricks, err = p.createBricks(namespace, name, cfg, gid)
+	if err != nil {
+		glog.Errorf("Creating bricks is failed: %s,%s", namespace, name)
+	}
+
+	if err == nil {
+		err = p.createGlusterVolume(bricks, cfg)
+	}
+
+	if err == nil {
+		epServiceName := dynamicEpSvcPrefix + name
+		epNamespace := namespace
+		dynamicHostIps := p.getClusterNodes(cfg)
+		endpoint, service, err = p.createEndpointService(epNamespace, epServiceName, dynamicHostIps, name)
+
+		if err != nil {
+			glog.Errorf("glusterfs: failed to create endpoint/service: %v", err)
+		} else {
+			glog.V(3).Infof("glusterfs: dynamic ep %v and svc : %v ", endpoint, service)
+			return &v1.GlusterfsVolumeSource{
+				EndpointsName: endpoint.Name,
+				Path:          cfg.VolumeName,
+				ReadOnly:      false,
+			}, nil
+		}
+	}
+
+	// XXX: Must not delete volume if creating volume is failed
```

Comment:
I'll delete this line and uncomment following code. Because these are derived from my old code.